### PR TITLE
bluetooth: host: hci_core: Safer checking of conn complete events

### PR DESCRIPTION
**Pull request by sean-madigan**
The spec only guarantees the connection complete event parameters 
are valid if the status is BT_HCI_ERR_SUCCESS. When the status is
not BT_HCI_ERR_SUCCESS, the host shall ignore the other parameters.

See Vol 4, Part E, 4.5 Command error handling:
```
If an error occurs for a command for which an HCI_Command_Complete event
is returned, the Return Parameters field may only contain some of the
return parameters specified for the command.
...
The above also applies to commands that have associated command specific
completion events with a Status parameter in their completion event, with
the exceptions shown in Table 4.1, Event Valid parameters
...
Event                           | Valid parameters
------------------------------------------------------------
LE_Connection_Complete          | none
LE_Enhanced_Connection_Complete | none
```

Refactor `le_legacy_conn_complete`, `le_enh_conn_complete_v2` and
`le_enh_conn_complete` to check and handle the status before handling
any other parameters.

An issue was seen where SDC returned event with status
`BT_HCI_ERR_UNKNOWN_CONN_ID`, but because adv_handle and sync_handle
were not invalid the event was not handled.

### CODE REVIEW

**Comment by olivier-le-sage** [comment]
> The spec only guarantees the status is valid if it is not BT_HCI_ERR_SUCCESS.

Suggest rephrasing to:

`The spec only guarantees the connection complete event parameters are valid if the status is BT_HCI_ERR_SUCCESS. When the status is not BT_HCI_ERR_SUCCESS, the host shall ignore the other parameters.`

File: subsys/bluetooth/host/hci_core.c [hunk 1]
```diff
@@ -1377,6 +1377,8 @@ static void update_conn(struct bt_conn *conn, const bt_addr_le_t *id_addr,
 
 void bt_hci_le_enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
 {
+	BT_ASSERT(evt->status == BT_HCI_ERR_SUCCESS);
```
Comment (Thalley) [inline subsys/bluetooth/host/hci_core.c:1380]: We should never assert (or rely on asserts) on parameters from 3rd party (we have to consider that the controller may be 3rd party here).

File: subsys/bluetooth/host/hci_core.c [hunk 2]
```diff
@@ -1654,9 +1624,44 @@ void bt_hci_le_enh_conn_complete_sync(struct bt_hci_evt_le_enh_conn_complete_v2
...
+
+	if (IS_ENABLED(CONFIG_BT_CENTRAL) &&
+			status == BT_HCI_ERR_UNKNOWN_CONN_ID) {
+		le_conn_complete_cancel(status);
+		bt_le_scan_update(false);
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CENTRAL) && IS_ENABLED(CONFIG_BT_PER_ADV_RSP) &&
+			status == BT_HCI_ERR_CONN_FAIL_TO_ESTAB) {
```
Comment (Thalley) [inline subsys/bluetooth/host/hci_core.c:1643]: Some indentation is odd here

**Review by Thalley — CHANGES_REQUESTED** [review]
Minor formatting comment, otherwise LGTM

File: subsys/bluetooth/host/hci_core.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (sean-madigan) [reply subsys/bluetooth/host/hci_core.c:1380]: Just saw `BT_ASSERT` elsewhere in the file and followed suit, am happy to use whichever is preferred

[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/host/hci_core.c:1380]: image
image

Looks like there is a clear preference for __ASSERT, but we don't have any guidelines or other to indicate what to use. I would assume that `BT_ASSERT` will be removed at some point. 

Either way is OK for this

**Review by Thalley — DISMISSED** [review]

File: subsys/bluetooth/host/hci_core.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (sean-madigan) [reply subsys/bluetooth/host/hci_core.c]: Updated to use `__ASSERT_NO_MSG`

**Review by Thalley — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

File: subsys/bluetooth/host/hci_core.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/bluetooth/host/hci_core.c]: > I would assume that BT_ASSERT will be removed at some point

I would not. Quite the contrary in fact: People in Zephyr don't like `CONFIG_ASSERT=y` in prod, so we should have our own, `default y` in Bluetooth. That's what `BT_ASSERT()` is.

**Review by jhedberg — APPROVED** [review]

**Review by jori-nordic — APPROVED** [review]

**Pull request merged**